### PR TITLE
Degree-day depassment date

### LINKED ISSUE #569: Implement degree-day depassment date indicator
#334 # Description

User interest for a calculation of date when the accumulated sum of degree-days reaches a user provided total.

E.g. the date when one reaches a sum of 100 degree-days (5 degC) starting from jan-01

I would suggest a 'generic' calculation allowing a single algorithm for all degree day types (Heating, cooling, growing etc etc). i.e. User provides parameter on the `op`  (<=, >=) as well as a `sum_threshold` (or similar), as well as all other params typically found in the degree day calculations already present.    

It would be interesting to see if there could be flexibility on whether the calculation is performed on daily tmin, tavg or tmax as well.

edit : Adding note that interested parties indicate that a `after_date` param would be ideal

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #569
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 
Adds `xclim.indices.degree_day_depassment_date`, which returns the day-of-year when the cumulative degree-day sum exceeds `sum_thresh`. The degree-days are computed using `thresh` and in the `op` direction. i.e.: if  `op = '>'`, `dd = (tas - thresh).clip(0)`
and inversely.  

This definition makes the cumulative degree-day sum a monotonically increasing function. Which raises the question, do we really need an `after_date`? It could be replaced by a `clip` on the result. But, it wasn't much work anyway.

This PR also extracts `xc.indices.generic.compare` from  `xc.indices.generic.threshold_count`. I thought I would need it, but then I didn't.

Finally, now all run_lengths with dates accept `date=None`, which result not performing that check. All hail polymorphism.

* **Does this PR introduce a breaking change?** 
No.

* **Other information**:

Indicators, doc and tests to come.

### CODE REVIEW

**Comment by aulemahal** [comment]
@tlogan2000  It this what we want?

**Comment by tlogan2000** [comment]
> @tlogan2000 It this what we want?

Will try to review today or tommorrow

File: xclim/indices/_threshold.py [hunk 1]
```diff
@@ -1365,3 +1366,31 @@ def tropical_nights(
...
+    thresh: str,
+    sum_thresh: str,
+    op: str = ">",
+    after_date: str = None,
+    freq: str = "YS",
+):
+    thresh = convert_units_to(thresh, "K")
+    tas = convert_units_to(tas, "K")
+    sum_thresh = convert_units_to(sum_thresh, "K days")
+    print(sum_thresh, thresh)
```
Comment (Zeitsperre) [inline xclim/indices/_threshold.py:1383]: This can be removed.

**Review by Zeitsperre — APPROVED** [review]
I _really_ like this. Very much what I had in mind. Well done, Pascal!
```
dd = degree_day_depassment_date(ds,thresh="5 degC", sum_thresh="500 degC days",freq="YS")
dd[0].plot()
plt.show()
```
Gives me exactly what I expect:
image

**Comment by aulemahal** [comment]
Woups didn't meant to do that.

**Review by Zeitsperre — APPROVED** [review]

**Comment by Zeitsperre** [comment]
@aulemahal Ignore the failing build. It'll be fixed in the next week or so.

File: xclim/indices/_threshold.py [hunk 2]
```diff
@@ -1365,3 +1366,74 @@ def tropical_nights(
...
+    if op in ["<", "<=", "lt", "le"]:
+        c = thresh - tas
+    elif op in [">", ">=", "gt", "ge"]:
+        c = tas - thresh
+
+    def _depassment_date(grp):
+        return rl.first_run_after_date(
+            grp.cumsum("time") > sum_thresh,
+            window=1,
+            date=after_date,
```
Comment (tlogan2000) [inline xclim/indices/_threshold.py:1436]: I don't think this is doing what is intended with the `after_date`.  My understanding from discussion with interested parties would be that value indicates the beginning of when the `cumsum()` should start accumulating values.  

So I think we would want to assign 'nan' or zeros to `grp`  before `after_date` value then let `rl.first_run_after_date()` run without the `date` input

**Review by tlogan2000 — CHANGES_REQUESTED** [review]
Looks good but I believe the `after_date` parameter is not doing exactly what we want. see comment in the code

File: xclim/indices/_threshold.py [hunk 3]
```diff
@@ -1365,3 +1366,74 @@ def tropical_nights(
...
+    tas : xarray.DataArray
+      Mean daily temperature.
+    thresh : str
+      Threshold temperature on which to base degree days evaluation.
+    sum_thresh : str
+      Threshold of the degree days sum, in "K days".
+    op : {">", "gt", "<", "lt", ">=", "ge", "<=", "le"}
+      If equivalent to '>', degree days are computed as `tas - thresh` and if
+      equivalent to '<', they are computed as `thresh - tas`.
+    after_date: str, optional
```
Comment (tlogan2000) [inline xclim/indices/_threshold.py:1396]: Modify after_date definition to 'start date of degree day accumulation - default is the beginning of resample period' or similar

File: xclim/indices/_threshold.py (hunk 2 above)
Comment (aulemahal) [reply xclim/indices/_threshold.py:1436]: Oh I see! In my approach I thought a user would choose the starting date by passing the appropriate freq, like "AS-JUL" for 1st of july, etc. But I can see how that is limiting.

Comment (tlogan2000) [reply xclim/indices/_threshold.py:1436]: That could work but I heard examples of some applications (Building codes or ??) that would  have super specific dates... "04-15" ... basically not always first of the month

**Review by tlogan2000 — APPROVED** [review]
Nice job.  FWIW seeing as TravisCI builds are not working I ran the full test suite on my VM

**Comment by Zeitsperre** [comment]
Builds should work now. Will remove this bypass when the new clisops is released.

**Pull request merged**